[DDP Logging] Log comm. hook in ddp logging

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52966 [DDP Logging] Log comm. hook in ddp logging**
* #52965 [DDP Logging] Log nccl_async_error_handling
* #52957 [Dist Debugality] Log key DDP metrics to stderr under debug mode.
* #52887 [Dist Debugability] Refactor and fix DDP model check during init

Logs registerd comm hook if there is one
Differential Revision: [D26709388](https://our.internmc.facebook.com/intern/diff/D26709388/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26709388/)!

rohan-varma committed:
[DDP Logging] Log comm. hook in ddp logging

Logs registerd comm hook if there is one, else logs
"builtin_allreduce"

Differential Revision: [D26709388](https://our.internmc.facebook.com/intern/diff/D26709388/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26709388/)!

[ghstack-poisoned]

SciPioneer commented on torch/lib/c10d/logger.hpp:
```diff
@@ -31,6 +31,8 @@ class Logger {
   // Get size of each bucket (Bytes).
   std::vector<int> get_bucket_sizes();
+  // Set comm. hook, if used
+  void set_comm_hook(const std::string &hook);
```
Nit: "std::string&" instead of "std::string &" for consistency.

SciPioneer commented on torch/lib/c10d/logger.cpp:
```diff
@@ -191,6 +198,13 @@ void Logger::set_runtime_stats_and_log() {
+
+  // Check if comm. hook has not been set, if so, set it as builtin allreduce.
+  if (!comm_hook_set_) {
+    ddp_logging_data_->comm_hook = kDefaultDDPComm;
```
If no hook is represented by an empty string, this added if statement can be removed.

SciPioneer commented on torch/lib/c10d/logger.hpp:
```diff
@@ -60,6 +62,8 @@ class Logger {
   // track the number of iterations when runtime stats are collected so far.
   long num_iterations_stats_recorded_ = 0;
+  // Track whether DDP comm. hook has been set or not
+  bool comm_hook_set_{false};
```
This field does not seem to be necessary, as its value can be determined by `ddp_logging_data_->comm_hook`.

SciPioneer commented on torch/lib/c10d/logger.cpp:
```diff
@@ -11,6 +11,8 @@ namespace {
 
 const int kMilliSecondToNanosSecond = 1000000;
 
+const char * kDefaultDDPComm = "builtin_allreduce";
```
This name can be confusing, since there is a built-in hook called AllReduce, but this value here actually means no hook, when allreduce is used w/o triggering a comm hook.

It will be better to use an empty string to represent the default value, indicating that no hook is registered.

rohan-varma replied:
Sounds good, will make the empty string the default/no comm hook registered value

wayi1 reviewed (commented):
Thanks for adding the log for comm hook!

zhaojuanmao commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -308,7 +308,12 @@ An enum-like class for built-in communication hooks: ``ALLREDUCE`` and ``FP16_CO
-          py::call_guard<py::gil_scoped_release>());
+          py::call_guard<py::gil_scoped_release>())
+        .def(
+            "set_comm_hook",
```
make it as private "_set_comm_hook_name"?

rohan-varma committed:
Update on "[DDP Logging] Log comm. hook in ddp logging"


Logs registerd comm hook if there is one, else logs
"builtin_allreduce"

Differential Revision: [D26709388](https://our.internmc.facebook.com/intern/diff/D26709388/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26709388/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP Logging] Log comm. hook in ddp logging"


Logs registerd comm hook if there is one, else logs
"builtin_allreduce"

Differential Revision: [D26709388](https://our.internmc.facebook.com/intern/diff/D26709388/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26709388/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP Logging] Log comm. hook in ddp logging"


Logs registerd comm hook if there is one, else logs
"builtin_allreduce"

Differential Revision: [D26709388](https://our.internmc.facebook.com/intern/diff/D26709388/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26709388/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP Logging] Log comm. hook in ddp logging"


Logs registerd comm hook if there is one, else logs
"builtin_allreduce"

Differential Revision: [D26709388](https://our.internmc.facebook.com/intern/diff/D26709388/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26709388/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP Logging] Log comm. hook in ddp logging"


Logs registerd comm hook if there is one, else logs
"builtin_allreduce"

Differential Revision: [D26709388](https://our.internmc.facebook.com/intern/diff/D26709388/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26709388/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP Logging] Log comm. hook in ddp logging"


Logs registerd comm hook if there is one, else logs
"builtin_allreduce"

Differential Revision: [D26709388](https://our.internmc.facebook.com/intern/diff/D26709388/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26709388/)!

[ghstack-poisoned]

ibro45 commented:
@rohan-varma do you think that this issue is connected to this PR or something you are familiar with? https://discuss.pytorch.org/t/distributed-1-8-0-logging-twice-in-a-single-process-same-code-works-properly-in-1-7-0/114103/6?u=ibro45
